Additional Chiller Part Load Curve Type

rongpengzhang opened the pull request:
@Introduction of the feature:
Chillers with variable speed drive (VSD Chillers) on compressors, reset control of entering condenser water temperature and leaving chilled water temperature can operate much more efficiently at part load conditions. An additional part-load performance curve type is proposed for such chillers in EnergyPlus. The new curve type is a modification to the existing curve and can model performance of VSD chillers more accurately.

@Summary of the changes:
(1) 2 existing CC files and 2 existing HH files are updated
(2) IDD file is modified
(3) 1 new IDF example file is added
(4) 10 existing IDF example files are updated

Myoldmopar commented:
OK, @rongpengzhang I pushed up some changes.  These include:
 - Updating transition source to accomodate the IDD changes.  This was a big burden, and I had to for some reason clean up a bunch of Transition source.  Anyway, the idfs are coming out clean and look great.  
 - Passing the relevant idfs through Transition to make sure they are clean and only show relevant diffs.  There are Sizing:System diffs in the field names only, which I think is because the idfs on develop were slightly out of date and hadn't been passed through Transition (looks like it went in with PR #4642)
 - Merging develop into this branch to ensure we are up to date to avoid any regression issues.

Let's wait and see what the robots say about this last commit (beefc61), and if everything looks good, we should be very close to merging.  Can you upload the doc changes (IORef, EngRef) to a new folder in [here](https://github.com/NREL/EnergyPlusBuildSupport/tree/master/docs/src/mods/for_2015_03_release)?

rongpengzhang commented:
Thanks, @Myoldmopar I can tell the transition updating is tough work. Good job!

I will upload the IORef and EngRef ASAP.

Myoldmopar commented:
@rongpengzhang Verified that the one failure should be cleaned up with the updated reference temperature.  This should be good to go with beefb3d and the doc changes over in BuildSupport.

@mjwitte This was proposed and approved a long time ago, so it's not worth a whole lot of scrutiny, but if you wouldn't mind taking a minute and glancing over, for example, the IDD changes, I would appreciate it.

Myoldmopar commented:
@rongpengzhang OK, things are looking good.  As expected, the cppcheck, unit test coverage, and integration test coverage all came back fully green.  And also as expected, the Ubuntu and Mac builds with regression tests are coming back with one diff, the idf we changed.  The WIndows machine is really busy, but I'll give it a little time this morning to get one result in.  This will give some more time to get the doc changes uploaded to BuildSupport anyway.

Myoldmopar commented:
Oh, and @rongpengzhang, in your pull request at the top, you @ tagged two random people: introduction and summary.  If you want titles, just use # or ## or ###, don't use the @ symbol.  ([Formatting tips here](https://help.github.com/articles/github-flavored-markdown/))

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -57223,17 +57223,41 @@ Chiller:Electric:ReformulatedEIR,
+       \note Calculated based on dT*, Tdev* and PLR
+       \note curve = a + b*(dT*) + c*(dT*)**2 + d*PLR + e*PLR**2 + f*(dT*)*PLR + g*(dT*)**3
+       \note + h*PLR**3 + i*(dT*)**2*PLR + j*(dT*)*PLR**2 + k*(dT*)**2*PLR**2 + l*(Tdev*)*PLR**3
+       \note Tdev = Leaving Chilled Water Temperature � Reference Chilled Water Temperature
```
@rongpengzhang There is a special character in this line - shows as a box on github.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -57223,17 +57223,41 @@ Chiller:Electric:ReformulatedEIR,
        \required-field
        \type object-list
        \object-list BicubicCurves
-       \object-list BiVariateTables
```
@rongpengzhang Do not delete BiVariateTables.  These are valid for the PLR curve type.
And add       \object-list MultivariateTables  which is valid for the lift case.  Appropriately dimensioned tables must always be acceptable as a curve substitute.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -57223,17 +57223,41 @@ Chiller:Electric:ReformulatedEIR,
-  A4 , \field Electric Input to Cooling Output Ratio Function of Part Load Ratio Curve Name
+  A4, \field Electric Input to Cooling Output Ratio Function of Part Load Ratio Curve Type
+       \note Two part-load curves are available. 
+       \note The previous one uses the leaving condenser water temperature.
```
@rongpengzhang Please change these notes to simply state there are two options, and then defined them, using the keyword and then what it means.  No need to talk about previous and new.  Something like:
\note Two curve types are available:
\note LeavingCondenserWaterTemperature - I'll just send a suggested IDD object by direct e-mail.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -57223,17 +57223,41 @@ Chiller:Electric:ReformulatedEIR,
+       \note water temperature deviates from the design conditions especially at high part-load conditions.
+       \required-field
+       \type choice
+       \key UseLeavingCondenserWaterTemperature
```
@rongpengzhang Subtle, but I would prefer to drop "Use" from these keywords, LeavingCondenserWaterTemperature or Lift

mjwitte commented on src/Transition/CreateNewIDFUsingRulesV8_3_0.f90:
```diff
@@ -321,29 +339,27 @@ SUBROUTINE CreateNewIDFUsingRules(EndOfFile,DiffOnly,InLfn,AskForInput,InputFile
+                CALL GetNewObjectDefInIDD(ObjectName,NwNumArgs,NwAorN,NwReqFld,NwObjMinFlds,NwFldNames,NwFldDefaults,NwFldUnits)
+                ! fields 1-9 aren't affected
+                OutArgs(1:9)=InArgs(1:9)
+                ! add a blank for the new curve type field
```
@Myoldmopar Any reason not to have transition insert the key choice here?  Much clearer to put "LeavingCondenserWaterTemperature" here.

Myoldmopar replied:
Sounds good to me!  I'll make the change once the IDD is updated with this as the key choice.

rongpengzhang commented:
@Myoldmopar  I'm happy that things are going well as what we expected. Thanks for the formatting tips, I didn't know there are better choices to make the title bold ~

@mjwitte Thanks a lot for the detailed IDD check and suggestion!
I will take care of the special character issue. 
Yes, "BiVariateTables" should be kept there. 
The updated IDD file will be uploaded very soon.

Myoldmopar commented:
@rongpengzhang CI looks great.  Once you are done with the IDD changes, I'll update the transition one more time per @mjwitte's comments, along with the idfs.  We'll have to let CI have one last pass to make sure the IDD changes don't accidentally mess up anything, but then merge.

rongpengzhang commented:
@Myoldmopar I have updated the IDD file taking @mjwitte 's comments. Can you let CI to run another test (cc and idf files are also changed due to idd updates)

mjwitte commented:
@rongpengzhang There's new warnings with HospitalLowEnergy?
@Myoldmopar I'm going to make a few more minor tweaks to the IDD - shouldn't impact any test runs.

mjwitte commented on src/EnergyPlus/ChillerReformulatedEIR.cc:
```diff
@@ -342,56 +355,77 @@ namespace ChillerReformulatedEIR {
-			if ( lAlphaFieldBlanks( 5 ) ) {
+			
+			//Check the type of part-load curves implemented: 1_LeavingCondenserWaterTemperature, 2_Lift    zrp_Aug2014
+			if ( SameString( PartLoadCurveType, "LeavingCondenserWaterTemperature" ) && SameString( GetCurveType( ElecReformEIRChiller( EIRChillerNum ).ChillerEIRFPLR ), "BICUBIC" ) ) {
```
@rongpengzhang Didn't notice these before.  These if blocks need to be expanded to include the valid table types.  Some components allow any form a curve as long as it has the same number of independent variables.  Do you feel strongly that this should only allow these specific curve forms?  e.g.  Should biquadratic be illegal, or just a warning, or completely acceptable?  @mmatts @Myoldmopar @EnergyArchmage Comments?

rongpengzhang replied:
@mjwitte 
It seems that the users are required to use "BICUBIC" , rather than other curve types, to describe this specific curve ("Electric Input to Cooling Output Ratio Function of Part Load Ratio Curve"). Details can be found in the following iDD sections. Please let me know if you think "warning" is better. Of course, table types need to be added if we keep using "severe error"

Chiller:Electric:ReformulatedEIR,
  A5 , \field Electric Input to Cooling Output Ratio Function of Part Load Ratio Curve Name
       \type object-list
       \object-list BicubicCurves
       \object-list BiVariateTables
       \object-list ChillerPartLoadWithLiftCurves
       \note two types of part-load curves are available: BicubicCurves or ChillerPartLoadWithLiftCurves
       \note Type 1: BicubicCurves; 
       \note Calculated based on LCT and PLR       
       \note curve = a + b*LCT + c*LCT**2 + d*PLR + e*PLR**2 + f*LCT*PLR + g*0 + h*PLR**3
       \note + i*0 + j*0
       \note PLR = part load ratio (cooling load/steady state capacity)
       \note LCT = leaving condenser fluid temperature(C)
       \note Type 2: ChillerPartLoadWithLiftCurves 
       \note Calculated based on dT*, Tdev* and PLR
       \note curve = a + b*(dT*) + c*(dT*)**2 + d*PLR + e*PLR**2 + f*(dT*)*PLR + g*(dT*)**3
       \note + h*PLR**3 + i*(dT*)**2*PLR + j*(dT*)*PLR**2 + k*(dT*)**2*PLR**2 + l*(Tdev*)*PLR**3

Myoldmopar replied:
I guess since the enforcement is actually on BiCubic curves _or_ tables, and if you enter BiCubic as the table type, this enforcement is fine.  Or am I missing something now?

Myoldmopar commented:
@mjwitte No, there are _fewer_ warnings in HospitalLowEnergy.  The lines in the [error file](http://energyplus.s3-website-us-east-1.amazonaws.com/b17c5be82c063b4d9ec0184c4c0b01a6c32de253-80d863916a8b9cd91fb9c6d4093f2eeff59cb98a/HospitalLowEnergy/x86_64-MacOS-10.9-clang/eplusout.err.diff.html) are now removed.

@mjwitte I'm guessing any two-independent-variable form is allowable, including tables, curves, and any shape.  Thus, it should be loosened up.

@rongpengzhang I will update the transition source per @mjwitte and check that in.  Then the curve type check in GetInput needs to be resolved.  _Hopefully_ that's that last issue.

mjwitte commented:
@Myoldmopar Whoops - sorry - still not used to the color green meaning deleted.  Error reduction is good.
@rongpengzhang I've committed suggested IDD changes.  Language about curve types may need to change slightly now from "Curve object type should be " to "Recommended curve types are:" ?

rongpengzhang commented:
@Myoldmopar I agree that this should be consistent with other similar cases in EnergyPlus. If the tradition is to consider it a "waning", it is fine with me.

rongpengzhang commented:
@mjwitte Thanks for the suggestion. I will do that

mjwitte commented:
@rongpengzhang @Myoldmopar Arrrgh.  I generally forget that tables can have a secondary curve type associated with them.  To me, the power of tables is to *not* use a specific curve form (i.e. Interpolation Type != EvaluateCurveToLimits).  @mmatts Is our expert on validation of curve types.  What are the other components doing?

Myoldmopar commented:
@rongpengzhang @mjwitte Updated transition source to write out the default value rather than a blank.  I'm not going to retransition those idfs.  They will get re-transitioned before release, plus I bet other idd changes are looming anyway.  So those fields will show up then.  I think @mmatts can confirm that the code is doing the right thing on the table, and then this should be OK.

Thanks for all the review time and conversation everyone.

mmatts commented:
@mjwitte It's been a while since I looked at the curves, but if I remember correctly we found out that the CurveManager is doing a good job of sorting out the right tables for the right types of curves. So long as the curves are referenced as they normally are throughout (which I think they are), I think this should be fine.

rongpengzhang commented:
@Myoldmopar  IDD files are updated again

Myoldmopar commented:
OK, CI should pick it up and _hopefully_ overnight it will come out all clean and we can merge it in _tomorrow_ morning.

rongpengzhang commented:
@Myoldmopar I have put the EngRef and IORef modifications into this folder: EnergyPlusBuildSupport/docs/src/mods/for_2015_03_release/ZRP-66516336-LBNLChillerCurve/

Myoldmopar commented:
Thanks @rongpengzhang for the doc changes, I'll look them over. I ran the new example file locally and found some _very_ minor messages (1 unused construction, etc.).  I'm not worried about that.

@mjwitte Does ExpandObjects write out reformulated EIR chillers?  If so, this will have to have an ExpandObjects update too, I guess.  Sorry I just thought about that, but 
 - [ ] ExpandObjects changes? 
is at the bottom of my review checklist... Moving it toward the top so it's not an afterthought.

mjwitte commented:
@Myoldmopar Reformulated EIR chiller is referenced, so there will be a few lines of code that need to change.

Myoldmopar commented:
Hooray!  OK, well I want this in today, so I will take a crack at making the changes in ExpandObjects.  I'll tag you for review if you can just glance over the ExpandObjects changes there.

mjwitte commented:
@Myoldmopar Search epfilter.f90 for 'reformulated'.  You just need to change some integer values to make sure they line up with the new object.  Also that "25" at the end of the long function call.

rongpengzhang commented:
@Myoldmopar We are almost there! Quite happy to see the delivery of my first EPlus product ~ Thanks Michaeland and Matt for all the review and support! Cheers ~

Myoldmopar commented:
@mjwitte OK, I updated the integers to account for the new field being inserted.  The new field is @ A4, which is overall argument # 9.  So all the node indices were incremented by 1, plus the last field index.  Should be good to go I would guess.  

With CI catching back up, if there are no more commits, we should have sufficient results to determine whether it is mergeable this afternoon.  Thanks for the help.

Oh, and since a separate PR went in this morning I'll merge develop right quick.
